tests/RevokePassportTokensTest.php

```diff
@@ -0,0 +1,292 @@
+
+namespace Jobilla\PassportRevoke\Tests;
+
+use Carbon\Carbon;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Jobilla\PassportRevoke\PassportRevokeServiceProvider;
+use Jobilla\PassportRevoke\RevokePassportTokens;
+use Laravel\Passport\Passport;
+use Laravel\Passport\PassportServiceProvider;
+use Laravel\Passport\Token;
+use Laravel\Passport\TokenRepository;
+use Mockery;
+use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
+use Orchestra\Testbench\TestCase;
+
+class RevokePassportTokensTest extends TestCase
+{
+    use MockeryPHPUnitIntegration, RefreshDatabase;
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        Carbon::setTestNow('2019-08-05 12:00:00');
+    }
+
+    public function test_it_can_revoke_a_specific_token()
+    {
+        Token::query()->forceCreate([
+            'id' => 'f8097432-b83e-4503-9c83-e3906655a58b',
+            'client_id' => 1,
+            'revoked' => false,
+            'expires_at' => Carbon::now()->addMonth(),
+        ]);
+
+        Token::query()->forceCreate([
+            'id' => 'dbca267d-acd8-4b6c-b770-e0035c31e2ca',
+            'client_id' => 1,
+            'revoked' => false,
+            'expires_at' => Carbon::now()->addMonth(),
+        ]);
+
+        $this->artisan(RevokePassportTokens::class, ['token' => 'f8097432-b83e-4503-9c83-e3906655a58b'])
+            ->assertExitCode(0)
+            ->run();
+        $this->assertEquals(1, Token::query()->where('revoked', false)->count());
+        $this->assertEquals(1, Token::query()->where('revoked', true)->count());
+    }
+
+    public function test_it_allows_revoking_all_tokens_if_no_conditions_are_passed()
+    {
+        Token::query()->insert([
+            [
+                'id' => 'b5b7ca3e-80e4-4b89-8905-8099e1a613cd',
+                'client_id' => 1,
+                'revoked' => 0,
+                'expires_at' => Carbon::now()->addMonth(),
+            ],
+            [
+                'id' => '469203f6-ba9d-4acf-88ed-c77dff94f4f6',
+                'client_id' => 1,
+                'revoked' => 0,
+                'expires_at' => Carbon::now()->addMonth(),
+            ],
+            [
+                'id' => '224f1318-9af1-4eb4-b3f3-0949b9e94fdd',
+                'client_id' => 2,
+                'revoked' => 0,
+                'expires_at' => Carbon::now()->addMonth(),
+            ],
+        ]);
+
+        $this->assertEquals(3, Token::query()->where('revoked', false)->count());
+
+        $this->artisan(RevokePassportTokens::class)
+            ->expectsQuestion(
+                'You did not provide any user, client or token. All Passport tokens will be revoked. Continue?',
+                true
+            )
+            ->assertExitCode(0)
+            ->run();
+
+        $this->assertEquals(3, Token::query()->where('revoked', true)->count());
+    }
+
+    public function test_it_does_not_revoke_all_tokens_if_the_user_does_not_confirm()
+    {
+        Token::query()->insert([
```

Comment:
True, that would increase the readability of this test a bit

Reply:
Note that there would still be duplication; what would disappear would be `id` and `expires_at`, so this would instead become:

```
$this->createTokens(
    [
        'client_id' => 1,
        'revoked' => 0,
    ],
    [
        'client_id' => 1,
        'revoked' => 0,
    ],
    [
        'client_id' => 2,
        'revoked' => 0,
    ],
```

I'm not sure it's _really_ worth it, but if you think it is, I'm happy to add it

Reply:
not sure why this would be necessary:
```
$this->createTokens(
    [
        'client_id' => 1,
        'revoked' => 0,
    ],
    [
        'client_id' => 1,
        'revoked' => 0,
    ],
    [
        'client_id' => 2,
        'revoked' => 0,
    ],
```

In all the 3 places you are inserting it 3 rows, the values you mention are unchanged, the only difference is that in one place you don't specify user_id, in the other 2 you do, but I guess it's not a problem to specify user ID for all of them, and you can even create all the 4 rows to make the set work for all of your tests. This will reduce the amount of duplications massively and improve readability IMO.

Reply:
I wouldn't want to create all the rows for all tests as that goes against the practice of creating the minimal required set for a test. I've added a method now that adds the id, expiry, and a default user ID. I'd like to keep the remainder explicit though so it's clear from just reading the test what test data is being used.

Reply:
Don't really get why you can't just create a base where you insert 3 rows, as now the only difference would be that the first 2 tests have single user id, the other one has 2 different user ids, having 2 different user ids wouldn't break the first 2 tests anyhow and doesn't add any extra row (so this meets the criteria of `creating the minimal required set for a test`). Now you can call `$this->createTokens();` for all the 3 tests. For the case where you have 4 rows, where 3 of them are exactly the same as for the previous 4 tests, you can call `$this->createTokens();` first and then create that extra record 
```
Token::query()->insert([
    'id' => '224f1318-9af1-4eb4-b3f3-0949b9e94fdd',
    'client_id' => 2,
    'revoked' => 0,
    'expires_at' => Carbon::now()->addMonth(),
    'user_id' => 2
]);
```

Reply:
The main problem is you can no longer directly see from *just reading the test* what the data looks like, you have to also look at the factory method, and then jump back.

Reply:
Well, you read this data only when the test starts to fail for some reason, and in that case you will anyway need to check the content of `createTokens` method, as now part of it is inline, and the other `common` part is outside of it. Don't see too much benefit, I must say. The reason in first place of asking to move `common` functionality to its own method is to make the code less heave and easy to read. Non of the targets were achieved by this solution. So might be better to revert to what it initially was...